### PR TITLE
Preprocessor matrix expansion

### LINKED ISSUE #32: Matrix expansion in config yaml
Per thoughts from @amoodie:
>Additionally, this could let us do cool things like config matrix expansions where something like:
```yaml
n_timesteps: 100

matrix:
  input_sand_fraction:
    - 0.4
    - 0.8
```
> produces two runs, each run for 100 timesteps. One run has sand frac 0.4, one has 0.6, and all other parameters are at default values. This could come much later, but could be a cool feature to have.

It would be nifty to have the input configuration yaml file allow for a matrix expansion. I will make this issue and flag it as an enhancement for the future.

### PR DESCRIPTION
**Pull request by amoodie**
Implement matrix expansion in the preprocessor.

For example: 

```yaml
out_dir: 'out_dir'
Length: 50.0
Width: 100.0
dx: 2.0
itermax: 1
Np_sed: 10
C0_percent: 0.1
N0_meters: 2.0
L0_meters: 1.0
save_eta_figs: True
save_figs_sequential: False
save_dt: 1

matrix:
  f_bedload: 
    - 0.5
    - 0.4
    - 0.2
  h0:
    - 1
    - 5
```

creates six jobs. E.g., the config of `job_000` uses `f_bedload: 0.5` and `h0: 1.0` and `job_001` uses `f_bedload: 0.5` and `h0: 5.0`:

`job_000`:
image

`job_001`:
image

## Questions

* Should we block the execution of jobs if the output folder already exists? I think this is a reasonable "safety" feature, but it may become an issue when we try to implement checkpointing, and we can also just remove it as a swim at your own risk type deal. I'm probably in favor of nixing this, but wanted to know what you thought.
* Similarly, I require you to specify `out_dir` in the input, but this isn't *strictly* necessary...The reason I did it is that I need this information in the `Preprocessor` to create the job folders, whereas normally this information is not processed until later, when all the yaml variables are pulled into the model itself. As I see it, the alternative to requiring the `out_dir` spec would be to read it from the `default.yml` file directly if it is not specified. Thoughts? 
* There is currently no limit to the number of jobs that can be expanded. Not sure if we should put a limit, or at least issue a warning? What is a reasonable limit?

### CODE REVIEW

File: pyDeltaRCM/preprocessor.py [hunk 1]
```diff
@@ -93,6 +165,11 @@ def construct_job_list(self):
         self.job_list = []
         if self._has_matrix:
             self.expand_yaml_matrix()
+            # loop the expanded jobs
+            for i in range(len(self.file_list)):
+                self.job_list.append(self._Job(self.file_list[i],
```
Comment (elbeejay) [inline pyDeltaRCM/preprocessor.py]: When an invalid value is provided as one of the matrix inputs, the initialization of the `_Job` class fails (since that initializes the `DeltaModel`). Since this throws an error, this stops any valid job in the matrix from running as well. 

Do you think it would be useful to instead have a `try-catch` statement here so that the valid jobs are still added to the job list and run, while the invalid jobs are "caught" and don't run?

File: pyDeltaRCM/shared_tools.py [hunk 2]
```diff
@@ -205,3 +205,19 @@ def _get_version():
...
+    file in the job output foler.
+    """
+    def _write_parameter_to_file(f, varname, varvalue):
+        """Write each line, formatted."""
+        f.write(varname + ': ' + str(varvalue) + '\n')
+
+    f = open(_path, "a")
+    for k in _config.keys():
+        _write_parameter_to_file(f, k, _config[k])
+    f.close()
```
Comment (elbeejay) [inline pyDeltaRCM/shared_tools.py]: Why is this in `shared_tools` as opposed to `preprocessor`?

**Review by elbeejay — COMMENTED** [review]
> Should we block the execution of jobs if the output folder already exists? I think this is a reasonable "safety" feature, but it may become an issue when we try to implement checkpointing, and we can also just remove it as a swim at your own risk type deal. I'm probably in favor of nixing this, but wanted to know what you thought.

The idea that one might accidentally overwrite a whole batch of jobs is pretty scary! I like the idea of having this 'safety' feature preventing that. We can probably change the structure of the `try-catch` loop to account for situations where the `resume_checkpoint` flag is set to `True` when we add that feature.

> Similarly, I require you to specify out_dir in the input, but this isn't strictly necessary...The reason I did it is that I need this information in the Preprocessor to create the job folders, whereas normally this information is not processed until later, when all the yaml variables are pulled into the model itself. As I see it, the alternative to requiring the out_dir spec would be to read it from the default.yml file directly if it is not specified. Thoughts?

I see what/why you've done it this way. I think it is fine as you've implemented it here. My take on this functionality overall is that it wouldn't be for someone who is using the code for the first time, and with the error you raise when `out_dir` isn't specified, it is pretty clear what has to be changed. I think this also ties in well with your previous point about overwriting jobs. If we fell back to the `default.yml` for the output directory then the overwriting case would probably get invoked more often too.

> There is currently no limit to the number of jobs that can be expanded. Not sure if we should put a limit, or at least issue a warning? What is a reasonable limit?

When the `verbose` flag is turned on right now, the number of jobs produced by the matrix expansion is printed to the console. I'm a bit wary about putting a limit or warning on the number of jobs; short runs on small domains can be much quicker than a 'standard' run, and long runs on huge domains can be much longer. Unless we wanted to try to estimate the "# of computations" from the `.yml` configuration before the model even runs, I'd say we leave the implementation here as is. 

#### Repeat values in `.yml` inside and outside of the matrix
In rare cases this actually becomes an issue, but in general it probably deserves at least a warning when it does happen so that runs with unintended parameters are not conducted. Basically if you specify an input parameter in the body of the `.yml` and then specify it **again** in the 'matrix', the jobs that are created only use the values provided in the matrix. In this way, someone might run the risk of overwriting an intended parameter value with whatever is defined in the matrix. An interesting note is that when the duplicated parameter is `out_dir`, then the matrix jobs are defined under `out_dir/job_001` etc and the `.yml` files sit there, however the output files and log files end up in directories that are defined in the 'matrix'. For example, this can be a problem with the matrix:

```
matrix:
  out_dir:
    - 'dir1'
    - 'dir2'

  dx:
    - 1.0
    - 2.0

  Length:
    - 20.0
    - 40.0
```

Creates 8 jobs, but will re-write output in the `dir1` and `dir2` directories as the `dx` and `Length` are changed. I guess the easy fix for this is to prohibit `out_dir` from being within the 'matrix' (or don't collect it).

Whichever way you end up going for the repeat `.yml` values, it might be worth adding a test to cover that scenario and verify expected output.

File: pyDeltaRCM/preprocessor.py (hunk 1 above)
Comment (amoodie) [reply pyDeltaRCM/preprocessor.py]: I'm not sure. I think as a user I would rather be told up front that the configuration is bad and that the suite of runs won't run as expected. Do you agree, or would you rather the runs that are able to run proceed? It's easy enough to implement the latter, but I think that could be confusing behavior...

I'll look at the specific error that is raised though, and maybe with a try-except for that specific error, we could provide a more helpful error message in the matrix expansion here.

Related: I have introduced a ton of `ValueError`s for the configuration...maybe we want to subclass the `Exception` with our own `YamlConfigurationError`. This would let us easily have a documentation page about how to handle each raised `YamlConfigurationError`. I'll probably open an issue about this, rather than address it now.

File: pyDeltaRCM/shared_tools.py (hunk 2 above)
Comment (amoodie) [reply pyDeltaRCM/shared_tools.py]: moved this to preprocessor as function.

**Comment by amoodie** [comment]
@elbeejay thanks for the feedback. I think I've addressed the cases you mentioned there with new input-checking and errors.

I [added a try-except for the bad YAML type configuration](https://github.com/amoodie/pyDeltaRCM_WMT/blob/080b6f47ac691e24d031aca632ebf9d26b67167a/pyDeltaRCM/preprocessor.py#L189) (and a test), but I still raise an error here. As I mentioned above, I think this is the more logical and expected behavior. I think this may not be totally necessary, as the original error message was pretty descriptive, but it definitely doesn't hurt anything.

The code on the exception is a bit ugly; I still don't know of the pretty way to write errors...

Once the tests pass, and I head back from you on whether you agree with this `try-except`-but-still-error-out approach, I'll be good to merge, if you approve :+1:

File: pyDeltaRCM/preprocessor.py (hunk 1 above)
Comment (elbeejay) [reply pyDeltaRCM/preprocessor.py]: > I think as a user I would rather be told up front that the configuration is bad and that the suite of runs won't run as expected.

Sure I can agree with this.

**Review by elbeejay — APPROVED** [review]
> @elbeejay thanks for the feedback. I think I've addressed the cases you mentioned there with new input-checking and errors.

Looks great, seems like the new error messages and unit tests cover at least the obvious edge cases that might come up. :+1:
